File: packages/sdk/src/administrator/components.ts
import BaseSDK from '../base-sdk';
import { QelosSDKOptions } from '../types';

export interface IComponent {
  _id?: string;
  tenant?: string;
  identifier: string;
  componentName: string;
  description?: string;
  content: string;
  requiredProps: {
    prop: string;
    type: string;
  }[];
  compiledContent?: {
    js: string;
    css: string;
  };
  created: Date;
  updated: Date;
}

type IComponentCreateParams = Pick<IComponent, 'identifier' | 'componentName' | 'content'> & {
  description?: string;
};

interface IBulkComponentResult {
  identifier: string;
  success: boolean;
  component?: IComponent;
  error?: string;
}

interface IBulkComponentResponse {
  total: number;
  succeeded: number;
  failed: number;
  results: IBulkComponentResult[];
}

type IComponentUpdateParams = Partial<Pick<IComponent, 'identifier' | 'componentName' | 'description' | 'content'>>;

export default class QlComponents extends BaseSDK {
  private relativePath = '/api/components';

  constructor(options: QelosSDKOptions) {
    super(options);
  }

  getComponent(componentId: string) {
    return this.callJsonApi<IComponent>(`${this.relativePath}/${componentId}`);
  }

  getList() {
    return this.callJsonApi<IComponent[]>(this.relativePath);
  }

  remove(componentId: string): Promise<IComponent> {
    return this.callJsonApi<IComponent>(`${this.relativePath}/${componentId}`, { method: 'delete' });
  }

  update(componentId: string, changes: IComponentUpdateParams): Promise<IComponent> {
    return this.callJsonApi<IComponent>(
      `${this.relativePath}/${componentId}`,
      {
        method: 'put',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify(changes)
      }
    );
  }

  create(component: IComponentCreateParams): Promise<IComponent> {
    return this.callJsonApi<IComponent>(this.relativePath, {
      method: 'post',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify(component)
    });
  }

  bulkCreateOrUpdate(components: IComponentCreateParams[]): Promise<IBulkComponentResponse> {
    return this.callJsonApi<IBulkComponentResponse>(`${this.relativePath}/bulk`, {
      method: 'post',
      headers: { 'content-type': 'application/json' },
      body: JSON.stringify({ components })
    });
  }
}

File: packages/python-sdk/qelos_sdk/administrator/components.py
from __future__ import annotations

import json
from typing import Any, Dict, List

from ..base_sdk import BaseSDK
from ..types import QelosSDKOptions


class QlComponents(BaseSDK):
    """Admin component management: CRUD operations on UI components."""

    _relative_path = "/api/components"

    def __init__(self, options: QelosSDKOptions) -> None:
        super().__init__(options)

    async def get_component(self, component_id: str) -> Dict[str, Any]:
        """Get a component by ID."""
        return await self.call_json_api(f"{self._relative_path}/{component_id}")

    async def get_list(self) -> List[Dict[str, Any]]:
        """List all components."""
        return await self.call_json_api(self._relative_path)

    async def remove(self, component_id: str) -> Dict[str, Any]:
        """Delete a component."""
        return await self.call_json_api(f"{self._relative_path}/{component_id}", method="DELETE")

    async def update(self, component_id: str, changes: Dict[str, Any]) -> Dict[str, Any]:
        """Update a component."""
        return await self.call_json_api(
            f"{self._relative_path}/{component_id}",
            method="PUT",
            headers={"content-type": "application/json"},
            body=json.dumps(changes),
        )

    async def create(self, component: Dict[str, Any]) -> Dict[str, Any]:
        """Create a new component.

        Args:
            component: Dict with ``identifier``, ``componentName``, ``content``,
                and optional ``description``.
        """
        return await self.call_json_api(
            self._relative_path,
            method="POST",
            headers={"content-type": "application/json"},
            body=json.dumps(component),
        )
